File: lib/qni/symbolic_state_renderer.rb
# frozen_string_literal: true

require 'json'
require 'open3'
require 'tmpdir'

module Qni
  # Invokes the Python symbolic helper for small symbolic state rendering.
  class SymbolicStateRenderer
    SUPPORTED_QUBIT_MESSAGE = 'symbolic run currently supports only 1-qubit and 2-qubit circuits'
    SETUP_MESSAGE = 'symbolic run requires SymPy runtime; run scripts/setup_symbolic_python.sh'
    HELPER_RELATIVE_PATH = '../../libexec/qni_symbolic_run.py'
    REPO_RUNTIME_RELATIVE_PATH = '../../.python-symbolic/bin/python'
    HELPER_ENV = { 'UV_CACHE_DIR' => File.join(Dir.tmpdir, 'qni-cli-uv-cache') }.freeze

    def initialize(circuit_hash, basis: nil)
      @circuit_hash = circuit_hash
      @basis = basis
    end

    def render
      render_with_format('text')
    end

    def render_latex_formula
      render_with_format('latex')
    end

    private

    def render_with_format(format)
      if %w[x y].include?(basis) && circuit_hash.fetch('qubits') != 1
        raise Simulator::Error, "symbolic #{basis}-basis run currently supports only 1-qubit circuits"
      end

      raise Simulator::Error, SUPPORTED_QUBIT_MESSAGE unless supported_qubit_count?

      args = ['--format', format]
      args.push('--basis', basis) if basis
      render_with_helpers(*args)
    end

    attr_reader :basis, :circuit_hash

    def supported_qubit_count?
      [1, 2].include?(circuit_hash.fetch('qubits'))
    end

    def helper_commands(*args)
      [
        [repo_runtime_path, helper_path, *args],
        ['python3', helper_path, *args],
        ['uv', 'run', '--quiet', '--with', 'sympy', 'python3', helper_path, *args]
      ]
    end

    def helper_path
      File.expand_path(HELPER_RELATIVE_PATH, __dir__)
    end

    def repo_runtime_path
      File.expand_path(REPO_RUNTIME_RELATIVE_PATH, __dir__)
    end

    def render_with_helpers(...)
      helper_commands(...).each do |command|
        output = render_with_helper(command)
        return output if output
      end

      raise Simulator::Error, SETUP_MESSAGE
    end

    def render_with_helper(command)
      stdout, stderr, status = run_helper(command)
      return stdout if status.success?

      handle_failed_helper(command, stderr, status)
    rescue Errno::ENOENT
      nil
    end

    def handle_failed_helper(command, stderr, status)
      helper_class = self.class
      return nil if helper_class.retryable_with_next_command?(stderr, command)

      raise Simulator::Error, helper_class.render_error_message(stderr, status)
    end

    def run_helper(command)
      stdout, stderr, status = Open3.capture3(
        HELPER_ENV,
        *command,
        stdin_data: JSON.generate(circuit_hash)
      )
      [stdout.strip, stderr, status]
    end

    class << self
      def retryable_with_next_command?(stderr, command)
        executable = command.first
        message = stderr.to_s
        (executable == 'uv' && message.include?('Failed to fetch:')) ||
          (executable == 'python3' && message.include?("No module named 'sympy'"))
      end

      def render_error_message(stderr, status)
        message = stderr.to_s.strip
        return message unless message.empty?

        "symbolic renderer failed with exit status #{status.exitstatus}"
      end
    end
  end
end

File: src/symbolic_state_renderer.ts
import { spawnSync } from 'node:child_process';
import { tmpdir } from 'node:os';
import path from 'node:path';

import type { CircuitData } from './circuit_file';

export class SymbolicStateRendererError extends Error {}

export type SymbolicOutputFormat = 'latex' | 'text';

export interface SymbolicStateRenderOptions {
  readonly basis?: string;
  readonly circuit: CircuitData;
  readonly env?: NodeJS.ProcessEnv;
  readonly format?: SymbolicOutputFormat;
  readonly projectRoot: string;
}

interface HelperCommand {
  readonly args: readonly string[];
  readonly command: string;
}

interface ResolvedSymbolicStateRenderOptions extends SymbolicStateRenderOptions {
  readonly format: SymbolicOutputFormat;
}

const SETUP_MESSAGE = 'symbolic run requires SymPy runtime; run scripts/setup_symbolic_python.sh';

export function renderSymbolicStateVector(options: SymbolicStateRenderOptions): string {
  validateSymbolicBasis({ basis: options.basis, qubits: options.circuit.qubits });

  return renderWithHelpers({
    ...options,
    format: options.format ?? 'text'
  });
}

function validateSymbolicBasis(options: { basis?: string; qubits: number }): void {
  if ((options.basis === 'x' || options.basis === 'y') && options.qubits !== 1) {
    throw new SymbolicStateRendererError(
      `symbolic ${options.basis}-basis run currently supports only 1-qubit circuits`
    );
  }

  if (options.basis === 'bell' && options.qubits !== 2) {
    throw new SymbolicStateRendererError('symbolic bell-basis run currently supports only 2-qubit circuits');
  }
}

function renderWithHelpers(options: ResolvedSymbolicStateRenderOptions): string {
  for (const command of helperCommands(options)) {
    const output = renderWithHelper(command, options);

    if (output !== undefined) {
      return output;
    }
  }

  throw new SymbolicStateRendererError(SETUP_MESSAGE);
}

function helperCommands(options: ResolvedSymbolicStateRenderOptions): HelperCommand[] {
  const helperPath = path.join(options.projectRoot, 'libexec', 'qni_symbolic_run.py');
  const args = [helperPath, '--format', options.format];

  if (options.basis !== undefined) {
    args.push('--basis', options.basis);
  }

  return [
    {
      args,
      command: path.join(options.projectRoot, '.python-symbolic', 'bin', 'python')
    },
    {
      args,
      command: 'python3'
    },
    {
      args: ['run', '--quiet', '--with', 'sympy', 'python3', ...args],
      command: 'uv'
    }
  ];
}

function renderWithHelper(command: HelperCommand, options: ResolvedSymbolicStateRenderOptions): string | undefined {
  const result = spawnSync(command.command, [...command.args], {
    encoding: 'utf8',
    env: {
      ...process.env,
      ...options.env,
      UV_CACHE_DIR: path.join(tmpdir(), 'qni-cli-uv-cache')
    },
    input: JSON.stringify(options.circuit),
    stdio: ['pipe', 'pipe', 'pipe']
  });

  if (result.error) {
    const error = result.error as NodeJS.ErrnoException;
    if (error.code === 'ENOENT') {
      return undefined;
    }

    if (error.code === 'EPIPE') {
      if (result.status === 0) {
        return result.stdout.trim();
      }

      if (retryableWithNextCommand(command.command, result.stderr)) {
        return undefined;
      }

      if (result.status !== null || result.stderr.trim() !== '') {
        throw new SymbolicStateRendererError(renderErrorMessage(result.stderr, result.status));
      }
    }

    throw new SymbolicStateRendererError(error.message);
  }

  if (result.status === 0) {
    return result.stdout.trim();
  }

  if (retryableWithNextCommand(command.command, result.stderr)) {
    return undefined;
  }

  throw new SymbolicStateRendererError(renderErrorMessage(result.stderr, result.status));
}

function retryableWithNextCommand(command: string, stderr: string): boolean {
  return (
    (command === 'uv' && stderr.includes('Failed to fetch:')) ||
    (command === 'python3' && stderr.includes("No module named 'sympy'"))
  );
}

function renderErrorMessage(stderr: string, status: number | null): string {
  const message = stderr.trim();
  return message === '' ? `symbolic renderer failed with exit status ${status ?? ''}` : message;
}
